src/App.jsx

```diff
@@ -27,23 +29,54 @@ import Team from "./views/Workspace/Sections/Team";
 
 
 const mapStateToProps = (state) => ({
+  playgrounds: state.playgrounds.playgrounds,
   })
   
 const mapDispatchToProps = (dispatch) => ({
-onUserSignedIn: user => dispatch(userSignedIn(user))
-})
+  onUserSignedIn: user => dispatch(userSignedIn(user)),
+  joinInitiative: initiativeId => dispatch(joinInitiative(initiativeId)),
+});
   
   
 class App extends React.Component {
 
-    signInHandler = (username, password) => {
+    joinAndOpenInitiative(user, history) {
+        const searchQuery = querySearch(history.location.search);
+
+        const url = searchQuery['from'];
+        const id = searchQuery['id'];
+
+        const playground = this.props.playgrounds
+          .find(({ id: playgroundId }) => playgroundId === id);
+
+        if (playground) {
+          const isUserAlreadyJoinedToInitiative = playground.volunteers
+            .find(({ userName }) => userName === user.username);
+
+          if (isUserAlreadyJoinedToInitiative) {
+            console.log('User is already joined to the initiative');
+
+            history.push(url);
+            return;
+          }
+        }
+
+        if (url && id) {
+            this.props.joinInitiative(id);
+            history.push(url);
+            return;
+        }
+
+        history.goBack();
+    }
+
+    signInHandler = (username, password, history) => {
```

Comment:
I have just pushed this logic back into the JSignin module, so you will have to rebase and move this logic there.

Reply:
Ok, thank you!